[css-grid] Split Grid and Masonry Implementation in resolveIntrinsicTrackSizes

stwrt opened the pull request:
#### edd20b55060cae41a9447146a25b92cf5e1d155d
<pre>
[css-grid] Split Grid and Masonry Implementation in resolveIntrinsicTrackSizes
<a href="https://bugs.webkit.org/show_bug.cgi?id=276831">https://bugs.webkit.org/show_bug.cgi?id=276831</a>
<a href="https://rdar.apple.com/problem/132103501">rdar://problem/132103501</a>

Reviewed by Sammy Gill.

Refactoring work to split up the implementations of resolveIntrinsicTrackSizes
and accumulateIntrinsicSizesForTrack.

Both of these implementations are constantly diverging based on optimizations I
have been working on. The Grid implementation is becoming more confusing to follow,
since it often does not need to care about the Masonry pieces.

Also, factored out handleInfinityGrowthLimit since it was just duplicating code that
was the exact same across implementations.

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrackMasonry):
(WebCore::GridTrackSizingAlgorithm::handleInfinityGrowthLimit):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizesMasonry):
(WebCore::GridTrackSizingAlgorithm::run):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: <a href="https://commits.webkit.org/281155@main">https://commits.webkit.org/281155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ded2ec42ba847f48318c4e168fbb25e0ae4ce7cd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58931 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38260 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62564 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9375 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61061 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45911 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9577 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47656 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6675 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60962 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35791 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50965 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28514 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32519 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8255 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8379 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54475 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8535 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64264 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8500 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54979 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2854 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50991 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55080 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13051 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2386 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34089 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35173 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36258 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34919 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1824,7 +1887,7 @@ void GridTrackSizingAlgorithm::run()
     // Step 2.
     if (!m_contentSizedTracksIndex.isEmpty())
-        resolveIntrinsicTrackSizes();
+        (m_renderGrid->isMasonry()) ? resolveIntrinsicTrackSizesMasonry() : resolveIntrinsicTrackSizes();
```
Since the non-masonry case is likely to be more common as of now I think it would be better to flip these

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1443,7 +1443,40 @@ bool GridTrackSizingAlgorithm::shouldExcludeGridItemForMasonryTrackSizing(const
+            itemsSortedByIncreasingSpan.append(GridItemWithSpan(*gridItem, span));
+    };
+
+    while (CheckedPtr<RenderBox> gridItem = iterator.nextGridItem())
```
I think it would be fine to change this just to `CheckedPtr` in this patch if you would like to do so

stwrt committed:
[css-grid] Split Grid and Masonry Implementation in resolveIntrinsicTrackSizes
https://bugs.webkit.org/show_bug.cgi?id=276831
rdar://problem/132103501

Reviewed by Sammy Gill.

Refactoring work to split up the implementations of resolveIntrinsicTrackSizes
and accumulateIntrinsicSizesForTrack.

Both of these implementations are constantly diverging based on optimizations I
have been working on. The Grid implementation is becoming more confusing to follow,
since it often does not need to care about the Masonry pieces.

Also, factored out handleInfinityGrowthLimit since it was just duplicating code that
was the exact same across implementations.

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrackMasonry):
(WebCore::GridTrackSizingAlgorithm::handleInfinityGrowthLimit):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizesMasonry):
(WebCore::GridTrackSizingAlgorithm::run):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: https://commits.webkit.org/281155@main
